lib/semantics/check-omp-structure.cc

```diff
@@ -0,0 +1,268 @@
+// Copyright (c) 2019, NVIDIA CORPORATION.  All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "check-omp-structure.h"
+#include "../parser/parse-tree.h"
+
+namespace Fortran::semantics {
+
+bool OmpStructureChecker::HasInvalidWorksharingNesting(
+    const parser::CharBlock &source, const OmpDirectiveSet &set) {
+  // set contains all the invalid closely nested directives
+  // for the given directive (`source` here)
+  if (CurrentDirectiveIsNested() && set.test(GetContext().currentDirectiveEnum)) {
+    context_.Say(source,
+        "A worksharing region may not be closely nested inside a "
+        "worksharing, explicit task, taskloop, critical, ordered, atomic, or "
+        "master region."_err_en_US);
+    return true;
+  }
+  return false;
+}
+
+void OmpStructureChecker::CheckAllowed(const OmpClause &type) {
+  if (ompContext_.empty()) {
+    return;
+  }
+  if ((!GetContext().allowedClauses.empty() &&
+          !GetContext().allowedClauses.test(type)) &&
```

Comment:
Is there a way to have a wrapper for the EnumSet::test() in OmpClauseSet which does the empty() check ?

Reply:
Yes, I can add a wrapper in EnumSet maybe. I will add it if these two functions are heavily used together in future development.